File: basic_dsa_four_hundred_six.java
/**
 * Basic DSA Problems - Queue Reconstruction by Height (LeetCode 406)
 * people[i]=[h,k]: k people in front with height>=h. Reconstruct queue.
 */
public class basic_dsa_four_hundred_six {

    // Problem: [h,k] - k = count of people in front with height >= h.

    /**
     * Sort by h descending, k ascending. Insert each at index k.
     * Time O(n^2), Space O(n)
     */
    public static int[][] reconstructQueue(int[][] people) {
        java.util.Arrays.sort(people, (a, b) -> a[0] != b[0] ? b[0] - a[0] : a[1] - b[1]);
        java.util.List<int[]> list = new java.util.ArrayList<>();
        for (int[] p : people) list.add(p[1], p);
        return list.toArray(new int[0][]);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Queue Reconstruction by Height");
        System.out.println("=".repeat(60));

        int[][] people = {{7,0},{4,4},{7,1},{5,0},{6,1},{5,2}};
        System.out.println("Result: " + java.util.Arrays.deepToString(reconstructQueue(people)));
    }
}

File: basic_dsa_four_hundred_six.cpp
/**
 * Basic DSA - Maximize Sum Of Array After K Negations (LeetCode 1005)
 * Choose an index and change A[i] to -A[i], exactly k times; maximize sum(A).
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

int largestSumAfterKNegations(std::vector<int>& nums, int k) {
    std::sort(nums.begin(), nums.end());
    int n = static_cast<int>(nums.size());
    for (int i = 0; i < n && k > 0 && nums[i] < 0; ++i) {
        nums[i] = -nums[i];
        --k;
    }
    std::sort(nums.begin(), nums.end());
    if (k % 2 == 1)
        nums[0] = -nums[0];
    long long s = 0;
    for (int x : nums)
        s += x;
    return static_cast<int>(s);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Maximize Sum Of Array After K Negations\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {4, 2, 3};
    std::cout << "[4,2,3], k=1 -> " << largestSumAfterKNegations(v1, 1) << "\n";

    std::vector<int> v2 = {3, -1, 0, 2};
    std::cout << "[3,-1,0,2], k=3 -> " << largestSumAfterKNegations(v2, 3) << "\n";
}
